cli/src/org/partiql/cli/prettyprintAST/PrettyPrinter.kt

```diff
@@ -0,0 +1,503 @@
+package org.partiql.cli.prettyprintAST
+
+import com.amazon.ion.system.IonSystemBuilder
+import org.partiql.lang.domains.PartiqlAst
+import org.partiql.lang.syntax.SqlParser
+import org.partiql.pig.runtime.SymbolPrimitive
+
+/**
+ * This class is used to pretty print PIG AST.
+ */
+class PrettyPrinter {
+    fun prettyPrintAST(query: String): String {
+        val ion = IonSystemBuilder.standard().build()
+        val ast = SqlParser(ion).parseAstStatement(query)
+
+        return prettyPrintAST(ast)
+    }
+
+    /**
+     * PIG AST is first transformed into a recursive tree structure, FeatureTree, then it is pretty printed.
+     */
+    fun prettyPrintAST(ast: PartiqlAst.Statement): String {
+        val featureTree = when (ast) {
+            is PartiqlAst.Statement.Query -> {
+                toFeatureTree(ast.expr)
+            }
+            is PartiqlAst.Statement.Dml -> TODO()
+            is PartiqlAst.Statement.Ddl -> TODO()
+            is PartiqlAst.Statement.Exec -> TODO()
+        }
+
+        return featureTree.convertToString()
+    }
+
+    private fun toFeatureTree(node: PartiqlAst.Expr, attrOfParent: String? = null): FeatureTree {
```

Comment:
Consider using the generated `org.partiql.lang.domains.PartiqlAst.Converter<T>`, which was designed for this kind of thing.

Reply:
I tried to use it, but I think it makes the code harder to understand. The reason is that only a part of PIG AST nodes have `Converter` interface. So if we convert some of the nodes using the interface, and some of nodes not using it, I think it does not look really good.